client/authority-discovery/src/interval.rs

```diff
@@ -0,0 +1,62 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use futures::stream::Stream;
+use futures::future::FutureExt;
+use futures::ready;
+use futures_timer::Delay;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+/// Exponentially increasing interval
+///
+/// Doubles interval duration on each tick until the configured maximum is reached.
+pub struct ExpIncInterval {
+	max: Duration,
+	next: Duration,
+	delay: Delay,
+}
+
+impl ExpIncInterval {
+	/// Create a new [`ExpIncInterval`].
+	pub fn new(start: Duration, max: Duration) -> Self {
+		let delay = Delay::new(start);
+		Self {
+			max,
+			next: start * 2,
+			delay,
+		}
+	}
+
+	/// Fast forward the exponentially increasing interval to the configured maximum.
+	pub fn set_to_max(&mut self) {
+		self.next = self.max;
+		self.delay = Delay::new(self.next);
+	}
+}
+
+impl Stream for ExpIncInterval {
+	type Item = ();
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+		ready!(self.delay.poll_unpin(cx));
+		self.delay = Delay::new(self.next);
```

Comment:
```suggestion
		self.delay = Delay::new(self.next);
		// Poll once to register the future in the current task.
		let _ = self.delay.poll_unpin(cx);
```

Reply:
Indeed this need to be done, otherwise the thing doesn't work at all.
However I'd suggest to put a loop around the entire body of `poll_next` instead of doing a dummy poll.

Reply:
Good catch @bkchr, thanks!

@tomaka can you expand on your idea of a loop around the entire body?

Reply:
Wait, no, you actually don't need this. (this code had passed my review the other day)

Since you return `Poll::Ready`, the caller *has* to call `poll_next` again, and this dummy registration isn't necessary.

To phrase it differently: the documentation of `poll_next` mentions that you pass a `Waker` that is woken up once the task is ready. In this case, you're returning `Ready`, so there's no need to register the `Waker`.

Reply:
Ahh yeah, I mixed Stream with Future. Sorry!